File: ios/Sasquatsh/Services/SessionsService.swift
import Foundation

protocol SessionsServiceProtocol {
    func getSessions(eventId: String) async throws -> EventSessionsResponse
    func registerForSession(sessionId: String) async throws
    func cancelSessionRegistration(sessionId: String) async throws
}

final class SessionsService: SessionsServiceProtocol {
    private let api: APIClient

    init(api: APIClient) {
        self.api = api
    }

    func getSessions(eventId: String) async throws -> EventSessionsResponse {
        let queryItems = [URLQueryItem(name: "eventId", value: eventId)]
        return try await api.get("sessions", queryItems: queryItems, authenticated: true)
    }

    func registerForSession(sessionId: String) async throws {
        let queryItems = [
            URLQueryItem(name: "action", value: "register"),
            URLQueryItem(name: "sessionId", value: sessionId)
        ]
        try await api.postVoid("sessions", queryItems: queryItems, authenticated: true)
    }

    func cancelSessionRegistration(sessionId: String) async throws {
        let queryItems = [URLQueryItem(name: "sessionId", value: sessionId)]
        try await api.deleteVoid("sessions", queryItems: queryItems, authenticated: true)
    }
}

File: android/app/src/main/java/com/sasquatsh/app/services/SessionsService.kt
package com.sasquatsh.app.services

import com.sasquatsh.app.models.*
import com.sasquatsh.app.services.api.SessionsApi
import com.squareup.moshi.Moshi
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class SessionsService @Inject constructor(
    private val sessionsApi: SessionsApi,
    private val moshi: Moshi
) {

    suspend fun getEventSessions(eventId: String): EventSessionsResponse {
        val response = sessionsApi.getSessions(eventId)
        if (!response.isSuccessful) throw Exception("Failed to load sessions")
        val json = moshi.adapter(Any::class.java).toJson(response.body())
        return moshi.adapter(EventSessionsResponse::class.java).fromJson(json)
            ?: throw Exception("Failed to parse sessions")
    }

    suspend fun registerForSession(sessionId: String) {
        val response = sessionsApi.registerForSession(sessionId = sessionId)
        if (!response.isSuccessful) throw Exception("Failed to register for session")
    }

    suspend fun unregisterFromSession(sessionId: String) {
        val response = sessionsApi.cancelSessionRegistration(sessionId = sessionId)
        if (!response.isSuccessful) throw Exception("Failed to unregister from session")
    }
}
